src/v/datalake/record_schema_resolver.cc

```diff
@@ -456,32 +456,59 @@ checked<std::vector<int32_t>, type_resolver::errc> compute_message_offsets(
 
 ss::future<checked<type_and_buf, type_resolver::errc>>
 latest_subject_schema_resolver::resolve_buf_type(std::optional<iobuf> b) const {
-    auto duration = std::chrono::duration_cast<ss::lowres_clock::duration>(
-      cache_duration_());
-    if (
-      latest_cached_schema_
-      && latest_cached_schema_->created_time + duration
-           > ss::lowres_clock::now()) {
-        co_return type_and_buf{
-          .type = std::make_optional(latest_cached_schema_->type.copy()),
-          .parsable_buf = std::move(b),
-        };
-    } else {
-        latest_cached_schema_ = std::nullopt;
+    const auto now = ss::lowres_clock::now();
+    const auto cache_ttl
+      = std::chrono::duration_cast<ss::lowres_clock::duration>(cache_ttl_());
+
+    if (schema_lookup_cache_.age(now) < cache_ttl) {
+        if (schema_lookup_cache_.entry().has_value()) {
+            co_return type_and_buf{
+              .type = std::make_optional(
+                schema_lookup_cache_.entry().value().copy()),
+              .parsable_buf = std::move(b),
+            };
+        } else {
+            if (datalake_log.is_enabled(ss::log_level::trace)) {
+                vlog(
+                  datalake_log.trace,
+                  "Using negative schema lookup cache for subject {}. "
+                  "Refreshing "
+                  "in {}s",
+                  subject_,
+                  std::chrono::duration_cast<std::chrono::seconds>(
+                    schema_lookup_cache_.time_until_expiry(now, cache_ttl))
+                    .count());
+            }
+            co_return schema_lookup_cache_.entry().error();
+        }
     }
+
+    // Expired cache. Reset it. We don't cache negative result for all error
+    // return paths because it is hard to reason whether methods are pure or
+    // not.
+    // Not impossible but not worth the complexity right now.
+    schema_lookup_cache_ = {};
+
+    auto effective_cache_timestamp = co_await sr_->sync(cache_ttl);
```

Comment:
nit: let's call this `last_sync_time` or something. IMO it's easier to follow along when the variable names convey _what_ this variable is (the last timestamp we synced at) rather than how it's used (the cache timestamp)